Service Internal Traffic Policy

maplain opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR implements the [Service Internal Traffic Policy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy)

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes/enhancements/issues/2086

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
Yes
```release-note
One new field "InternalTrafficPolicy" in Service is added.
It specifies if the cluster internal traffic should be routed to all endpoints or node-local endpoints only.
"Cluster" routes internal traffic to a Service to all endpoints.
"Local" routes traffic to node-local endpoints only, and traffic is dropped if no node-local endpoints are ready.
The default value is "Cluster".
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- [KEP]: <https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2086-service-internal-traffic-policy>

```docs
TBD
```

Commit:
Implements Service Internal Traffic Policy

1. Add API definitions;
2. Add feature gate and drops the field when feature gate is not on;
3. Set default values for the field;
4. Add API Validation
5. add kube-proxy iptables and ipvs implementations
6. add tests

andrewsykim commented:
/triage accepted
/milestone v1.21
/assign

Thanks for starting this @maplain!

andrewsykim commented on pkg/apis/core/types.go:
```diff
@@ -3674,6 +3693,16 @@ type ServiceSpec struct {
+	// "PreferLocal" will route internal traffic to node-local endpoints if
+	// one exists, otherwise it will fallback to the same behavior as
+	// "Cluster". "Local" routes traffic to node-local endpoints only,
+	// traffic is dropped if no node-local endpoints are ready.
```
add a `+optional` similar to `ExternalTrafficPolicy` above.

andrewsykim commented on pkg/apis/core/types.go:
```diff
@@ -3674,6 +3693,16 @@ type ServiceSpec struct {
+	// one exists, otherwise it will fallback to the same behavior as
+	// "Cluster". "Local" routes traffic to node-local endpoints only,
+	// traffic is dropped if no node-local endpoints are ready.
+	InternalTrafficPolicy ServiceInternalTrafficPolicyType
```
We need to also update the v1 types: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L4032-L4253

And run `make update` to update generated code.

andrewsykim commented:
I know this PR is a WIP, but some comments for what is still required:
1. We need a `ServiceInternalTrafficPolicy` feature gate in https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go that is alpha status and off by default.
2. We need to drop the new field in the Service REST code when the feature gate is off. See `spec.allocateLoadBalancerNodePorts` as one example: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/strategy.go#L185-L190
3. We need to set the default value to `Cluster` if the feature gate is enabeld, see `SetDefaults_Service`: https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L97

andrewsykim commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4163,6 +4182,17 @@ type ServiceSpec struct {
+	// "Cluster". "Local" routes traffic to node-local endpoints only,
+	// traffic is dropped if no node-local endpoints are ready.
+	// +optional
+	InternalTrafficPolicy ServiceInternalTrafficPolicyType
```
The published v1 types need json tags (see `externalTrafficPolicy` above for format). Need to use an unused protobuf number though (21 is the next one I believe)

andrewsykim replied:
This also needs to be a string pointer

andrewsykim commented on pkg/apis/core/types.go:
```diff
@@ -3674,6 +3693,17 @@ type ServiceSpec struct {
+	// "Cluster". "Local" routes traffic to node-local endpoints only,
+	// traffic is dropped if no node-local endpoints are ready.
+	// +optional
+	InternalTrafficPolicy ServiceInternalTrafficPolicyType
```
This actually needs to be a string pointer, so we can drop the field when disabled.

maplain replied:
would being inconsistent with `externalTrafficPolicy` a concern?

andrewsykim replied:
I don't think so, `externalTrafficPolicy` pre-dates the requirement that new string fields have to be pointers (so they can be nil-able and entirely removed from API responses).

andrewsykim replied:
(from a user dealing with json/yaml, there should be no functional difference)

andrewsykim commented on pkg/registry/core/service/strategy.go:
```diff
@@ -251,6 +258,17 @@ func loadBalancerPortsInUse(svc *api.Service) bool {
+		return false
+	}
+	switch svc.Spec.InternalTrafficPolicy {
+	case api.ServiceInternalTrafficPolicyCluster, api.ServiceInternalTrafficPolicyPreferLocal, api.ServiceInternalTrafficPolicyLocal:
```
I don't think we need to actually check the values, this can probably just be:
```
if svc == nil {
    return false
}

return svc.Spec.InternalTrafficPolicy != nil 
```
(after we updated InternalTrafficPolicy to string pointer)

andrewsykim commented:
/ok-to-test

maplain commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

maplain commented:
@andrewsykim tested locally in kind in both iptables and ipvs modes
1. When it's Local mode, only query on the non-"empty" nodes goes through;
2. When it's PreferLocal, query on the non-"empty" nodes only goes to the node local endpoints and query on the "empty" nodes is round-robined among all endpoints;
3. When it's All, it's the same behavior as before;

andrewsykim commented on pkg/apis/core/v1/defaults.go:
```diff
@@ -131,6 +131,11 @@ func SetDefaults_Service(obj *v1.Service) {
 		obj.Spec.ExternalTrafficPolicy = v1.ServiceExternalTrafficPolicyTypeCluster
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) && obj.Spec.InternalTrafficPolicy == nil {
```
Can we add a unit test for this in defaults_test.go?

maplain replied:
done

andrewsykim commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1030,6 +1030,15 @@ func (proxier *Proxier) syncProxyRules() {
+		// following are true:
+		// 1. InternalTrafficPolicy is PreferLocal or Local
+		// 2. ServiceInternalTrafficPolicy feature gate is on
+		if svcInfo.PreferNodeLocalEndpoints() && utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
Can we add a unit test for the iptable rules? I don't think we have to test the iptables datapath (that would be e2e tests) but testing the expected rules is probably good. See `TestEndpointSliceE2E` for an example.

maplain replied:
done

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -2041,6 +2042,14 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
+	// following are true:
+	// 1. InternalTrafficPolicy is PreferLocal or Local
+	// 2. ServiceInternalTrafficPolicy feature gate is on
+	if preferNodeLocalEndpoints && utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
Can we add a unit test here?

maplain replied:
done

andrewsykim commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4353,6 +4353,19 @@ func ValidateService(service *core.Service) field.ErrorList {
+
+	// internal traffic policy field
+	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
+		if service.Spec.InternalTrafficPolicy != nil &&
```
Can we put this in a function similar to `validateServiceExternalTrafficFieldsValue`?

andrewsykim commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4353,6 +4353,19 @@ func ValidateService(service *core.Service) field.ErrorList {
 	allErrs = append(allErrs, validateServiceExternalTrafficFieldsValue(service)...)
+
+	// internal traffic policy field
+	if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
I think we can rmove the feature gate check here as long as we check for `nil` in the validation check. Since disabled feature gate means the field is always nil.

andrewsykim commented on pkg/proxy/topology.go:
```diff
@@ -78,3 +78,37 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
+// array of Endpoints; otherwise, only return the node local endpoints.
+//
+// If InternalTrafficPolicy is Local, only return the node local endpoints.
+func FilterLocalEndpoint(internalTrafficPolicy *v1.ServiceInternalTrafficPolicyType, endpoints []Endpoint) []Endpoint {
```
Can we add a unit test here similar to `TestFilterTopologyEndpoint`?

maplain replied:
done

andrewsykim commented on pkg/proxy/service.go:
```diff
@@ -119,6 +121,16 @@ func (info *BaseServiceInfo) OnlyNodeLocalEndpoints() bool {
 
+// PreferNodeLocalEndpoints is part of ServicePort interface
+func (info *BaseServiceInfo) PreferNodeLocalEndpoints() bool {
+	return info.preferNodeLocalEndpoints
```
What do you think about removing `PrefersLocalTraffic` in apiservice package and having this return true or false based on `info.internalTrafficPolicy`? Just one less field to carry around in `BaseServiceInfo`.

maplain replied:
kk. just trying to be aligned with existing behavior for externalTrafficPolicy: `RequestsOnlyLocalTraffic`. I could refactor it if you prefer directly making a decision here.

maplain replied:
resolving it now to favor consistency. feel free to reopen it if you want me to change it.

andrewsykim commented:
PR is looking great, thanks @maplain! Most comments are just around unit testing.

maplain commented on pkg/api/v1/service/util.go:
```diff
@@ -76,6 +76,18 @@ func RequestsOnlyLocalTraffic(service *v1.Service) bool {
 }
 
+// PrefersLocalTraffic checks if service prefers Node Local traffic
+func PrefersLocalTraffic(service *v1.Service) bool {
```
done

maplain commented:
/retest

maplain commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

maplain commented:
/test pull-kubernetes-bazel-test

andrewsykim commented on pkg/apis/core/types.go:
```diff
@@ -3465,6 +3465,24 @@ const (
+
+	// ServiceInternalTrafficPolicyPreferLocal routes to node-local
+	// endpoints if it exists, otherwise fall backs to all
+	// endpoints
```
nit: this can probably go up to the previous line, no need to line break here

andrewsykim commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -2052,3 +2052,58 @@ func TestSetDefaultServiceIPFamilyPolicy(t *testing.T) {
+	}
+	for _, test := range testCases {
+		t.Run(test.name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceInternalTrafficPolicy, true)()
```
include the feature gate bool in the test table and set it here so we can also test behavior when the gate is off

andrewsykim commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4422,6 +4426,22 @@ func validateServiceExternalTrafficFieldsValue(service *core.Service) field.Erro
+		*service.Spec.InternalTrafficPolicy != core.ServiceInternalTrafficPolicyPreferLocal &&
+		*service.Spec.InternalTrafficPolicy != core.ServiceInternalTrafficPolicyLocal {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("internalTrafficPolicy"), service.Spec.InternalTrafficPolicy,
+			fmt.Sprintf("InternalTrafficPolicy must be empty, %v, %v or %v", core.ServiceInternalTrafficPolicyAll, core.ServiceInternalTrafficPolicyPreferLocal, core.ServiceInternalTrafficPolicyLocal)))
```
nit: InternalTrafficPolicy -> internalTrafficPolicy

andrewsykim commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4422,6 +4426,22 @@ func validateServiceExternalTrafficFieldsValue(service *core.Service) field.Erro
+		*service.Spec.InternalTrafficPolicy != core.ServiceInternalTrafficPolicyPreferLocal &&
+		*service.Spec.InternalTrafficPolicy != core.ServiceInternalTrafficPolicyLocal {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("internalTrafficPolicy"), service.Spec.InternalTrafficPolicy,
+			fmt.Sprintf("InternalTrafficPolicy must be empty, %v, %v or %v", core.ServiceInternalTrafficPolicyAll, core.ServiceInternalTrafficPolicyPreferLocal, core.ServiceInternalTrafficPolicyLocal)))
```
nit: this might be a bit cleaner if we use a string set with all the supported values, see usage of `supportedServiceType` as an example:

https://github.com/kubernetes/kubernetes/blob/c5b07f09671f7a0b4d458eb4da64a0b2195bbecf/pkg/apis/core/validation/validation.go#L4247-L4251

andrewsykim replied:
https://github.com/kubernetes/kubernetes/blob/c5b07f09671f7a0b4d458eb4da64a0b2195bbecf/pkg/apis/core/validation/validation.go#L4150-L4151

andrewsykim commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1020,6 +1020,13 @@ func (proxier *Proxier) syncProxyRules() {
+		// following are true:
+		// 1. InternalTrafficPolicy is PreferLocal or Local
+		// 2. ServiceInternalTrafficPolicy feature gate is on
+		if svcInfo.PreferNodeLocalEndpoints() && utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
nit: i would flip the order here:
```
if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) && svcInfo.PreferNodeLocalEndpoints() {
```

andrewsykim commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -26,6 +26,7 @@ import (
 	"testing"
 	"time"
 
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
```
nit: this can probably be grouped with other imports below

andrewsykim commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -2687,3 +2690,288 @@ COMMIT
+:KUBE-SERVICES - [0:0]
+:KUBE-EXTERNAL-SERVICES - [0:0]
+:KUBE-FORWARD - [0:0]
+-A KUBE-SERVICES -m comment --comment "ns1/svc1 has no endpoints" -m tcp -p tcp -d 172.20.1.1/32 --dport 0 -j REJECT
```
does `--dport 0` mean all ports here?

maplain replied:
weird. not sure why this comment doesn't show up in the `Conversations` tab.

This is because `TargetPort` instead of `Port` is used in the Service. Fixed.

andrewsykim commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -2687,3 +2690,288 @@ COMMIT
+	}
+
+	for _, tc := range testCases {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceInternalTrafficPolicy, true)()
```
Might be worthwhile to have a feature gate toggle in the test table so we can test that nothing changes when the feature gate is off

robscott replied:
+1

maplain replied:
added one testcase for feature gate being off to ensure it's the same rules as `internalTrafficPolicy=Cluster`

andrewsykim commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -2029,6 +2030,14 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
+	// following are true:
+	// 1. InternalTrafficPolicy is PreferLocal or Local
+	// 2. ServiceInternalTrafficPolicy feature gate is on
+	if preferNodeLocalEndpoints && utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
nit: flip the order:
```
if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) && preferNodeLocalEndpoints {
```

andrewsykim commented on pkg/proxy/topology.go:
```diff
@@ -78,3 +78,40 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
+// FilterLocalEndpoint should never be called when InternalTrafficPolicy is All.
+// If it is, all endpoints will be directly returned
+func FilterLocalEndpoint(internalTrafficPolicy *v1.ServiceInternalTrafficPolicyType, endpoints []Endpoint) []Endpoint {
+	// Should never happen
```
this comment is probably not needed

andrewsykim commented on pkg/proxy/topology_test.go:
```diff
@@ -476,3 +476,138 @@ func TestFilterTopologyEndpoint(t *testing.T) {
+		expected              []endpoint
+	}{
+		{
+			// Case[0]: no internalTrafficPolicy set with empty
```
really minor nit: we probably don't need these comments since we have the same description in the test names.

andrewsykim commented on pkg/proxy/topology_test.go:
```diff
@@ -476,3 +476,138 @@ func TestFilterTopologyEndpoint(t *testing.T) {
+	preferLocal := v1.ServiceInternalTrafficPolicyPreferLocal
+
+	testCases := []struct {
+		Name                  string
```
nit: this can just be `name`

andrewsykim commented on pkg/proxy/topology_test.go:
```diff
@@ -476,3 +476,138 @@ func TestFilterTopologyEndpoint(t *testing.T) {
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			m := make(map[Endpoint]endpoint)
+			endpoints := []Endpoint{}
```
Would it be cleaner to just have `[]Endpoint{}` as both the input and expected output in the test table?

andrewsykim replied:
This way we don't have to copy into another map

andrewsykim commented on pkg/registry/core/service/strategy.go:
```diff
@@ -173,6 +173,13 @@ func dropServiceDisabledFields(newSvc *api.Service, oldSvc *api.Service) {
 	}
+
+	// Clear InternalTrafficPolicy if not enabled
+	if !utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) {
```
Can probably add a testcase for this in TestDropDisabledField

andrewsykim reviewed (commented):
Mostly just nits at this point

maplain commented:
/test pull-kubernetes-bazel-test

maplain commented:
/test pull-kubernetes-bazel-test

maplain commented:
/test pull-kubernetes-bazel-test

robscott commented:
/cc

maplain commented:
@andrewsykim @robscott Could you please take a look?

robscott commented on pkg/api/v1/service/util.go:
```diff
@@ -76,6 +76,19 @@ func RequestsOnlyLocalTraffic(service *v1.Service) bool {
 
+// RequestsOnlyLocalTrafficForInternal checks if service prefers Node Local
+// endpoints for internal traffic
+func RequestsOnlyLocalTrafficForInternal(service *v1.Service) bool {
+	if service.Spec.InternalTrafficPolicy == nil {
+		return false
+	}
+	switch *service.Spec.InternalTrafficPolicy {
+	case v1.ServiceInternalTrafficPolicyLocal:
+		return true
+	}
+	return false
+}
```
Although this likely came from when there were more policies, this feels a bit overcomplicated now.

maplain replied:
This is to be aligned with externalTrafficPolicy: https://github.com/kubernetes/kubernetes/pull/96600/files/d0f3fd726b6a470c56d70e904db2407f9f345ae3#diff-514495aa2eda84bb5aed2784b8f2314e75b2f2861a369ef14c4a9885382f68eeR70

But I'm open to changes.

maplain replied:
updated

robscott commented on pkg/api/v1/service/util_test.go:
```diff
@@ -214,3 +214,30 @@ func TestNeedsHealthCheck(t *testing.T) {
 }
+
+func TestRequestsOnlyLocalTrafficForInternal(t *testing.T) {
+	checkRequestsOnlyLocalTrafficForInternal := func(expected bool, service *v1.Service) {
```
Nit: I think I'd find this easier to understand if it was using the test tables approach that's fairly common in k/k.

maplain replied:
That makes sense to me. The rest of the tests in this file are all in this format though..

robscott commented on pkg/apis/core/v1/defaults_test.go:
```diff
@@ -1798,3 +1801,64 @@ func TestSetDefaultEnableServiceLinks(t *testing.T) {
+			featureGateOn: true,
+		},
+		{
+			name:                          "must not set default internalTrafficPolicy",
```
tiny nit:
```suggestion
			name:                          "must not set default internalTrafficPolicy when gate is disabled",
```

maplain replied:
updated

robscott commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10059,6 +10059,7 @@ func TestValidatePodEphemeralContainersUpdate(t *testing.T) {
 
 func TestValidateServiceCreate(t *testing.T) {
 	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceTopology, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ServiceInternalTrafficPolicy, true)()
```
Wouldn't this validation happen even if the feature gate wasn't enabled?

maplain replied:
you're right. removed this line

robscott commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1025,10 +1025,17 @@ func (proxier *Proxier) syncProxyRules() {
 		// 3. EndpointSlice is not enabled (service topology depends on endpoint slice
 		// to get topology information).
-		if !svcInfo.OnlyNodeLocalEndpoints() && utilfeature.DefaultFeatureGate.Enabled(features.ServiceTopology) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceProxying) {
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceTopology) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceProxying) {
```
Why is the `onlyNodeLocalEndpoints()` check being removed here?

maplain replied:
fixed

robscott commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1025,10 +1025,17 @@ func (proxier *Proxier) syncProxyRules() {
 			allEndpoints = proxy.FilterTopologyEndpoint(proxier.nodeLabels, svcInfo.TopologyKeys(), allEndpoints)
 		}
 
+		// Service InternalTrafficPolicy is only enabled when all of the
+		// following are true:
+		// 1. InternalTrafficPolicy is Local
+		// 2. ServiceInternalTrafficPolicy feature gate is on
+		if utilfeature.DefaultFeatureGate.Enabled(features.ServiceInternalTrafficPolicy) && svcInfo.OnlyNodeLocalEndpointsForInternal() {
+			allEndpoints = proxy.FilterLocalEndpoint(svcInfo.InternalTrafficPolicy(), allEndpoints)
+		}
```
One of the things I'm doing in the topology PR is pushing all of this logic into proxy so it's not repeated in each proxier. That would turn this into a call like this:

```
allEndpoints = proxy.FilterEndpoints(allEndpoints, svcInfo, proxier.nodeLabels)
```

That function then determines how to filter endpoints based on a combination of `svcInfo` and feature gates. I don't think that's strictly required for this PR, but wherever possible, I prefer pulling shared logic out of proxier implementations and into shared code.
https://github.com/kubernetes/kubernetes/pull/99522/files#diff-2ccafd63d57233fd0697b9852514604ab880e29aeb073035765ad7541cad1b35R1027

maplain replied:
This is definitely better and cleaner! I'll defer this change to your PR.

robscott commented on pkg/proxy/topology.go:
```diff
@@ -78,3 +78,29 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
+		return endpoints
+	}
+
+	var filteredEndpoint []Endpoint
```
Tiny nit
```suggestion
	var filteredEndpoints []Endpoint
```

maplain replied:
fixed

robscott reviewed (commented):
Thanks for the work on this, just a few small nits, but otherwise LGTM.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3684,6 +3697,14 @@ type ServiceSpec struct {
+	// should be routed to all endpoints or node-local endpoints only
+	// "Cluster" routes internal traffic to a Service to all endpoints.
+	// "Local" routes traffic to node-local endpoints only, traffic is
+	// dropped if no node-local endpoints are ready.
```
Also specify the default

thockin commented on pkg/proxy/service.go:
```diff
@@ -39,18 +39,20 @@ import (
+	externalIPs                       []string
+	loadBalancerSourceRanges          []string
+	healthCheckNodePort               int
+	onlyNodeLocalEndpoints            bool
```
as a cleanup (these names are pretty bad): how about `nodeLocalExternal` and `nodeLocalInternal` ?

thockin replied:
the rename could be a new commit

maplain replied:
done

thockin commented on pkg/proxy/service.go:
```diff
@@ -119,6 +121,16 @@ func (info *BaseServiceInfo) OnlyNodeLocalEndpoints() bool {
 }
 
+// OnlyNodeLocalEndpointsForInternal is part of ServicePort interface
+func (info *BaseServiceInfo) OnlyNodeLocalEndpointsForInternal() bool {
```
We could also rename these accessor methods like `NodeLocalInternal` and `NodeLocalExternal`

maplain replied:
done

thockin commented on pkg/proxy/service.go:
```diff
@@ -129,6 +141,10 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
 		onlyNodeLocalEndpoints = true
 	}
+	onlyNodeLocalEndpointsForInternal := false
+	if apiservice.RequestsOnlyLocalTrafficForInternal(service) {
```
honestly, these helper methods are pointless but you get credit for following the pattern :)

thockin commented on pkg/proxy/topology.go:
```diff
@@ -69,12 +69,38 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
+//
+// If InternalTrafficPolicy is Local, only return the node local endpoints.
+//
+// FilterLocalEndpoint should never be called when InternalTrafficPolicy is All.
```
ss/All/Cluster

thockin replied:
but the code handles this case fine?  Just remove comment?

maplain replied:
removed comment

thockin commented on pkg/proxy/service.go:
```diff
@@ -129,6 +141,10 @@ func (sct *ServiceChangeTracker) newBaseServiceInfo(port *v1.ServicePort, servic
 		onlyNodeLocalEndpoints = true
 	}
+	onlyNodeLocalEndpointsForInternal := false
+	if apiservice.RequestsOnlyLocalTrafficForInternal(service) {
```
this needs to check the gate - it's possible the gate may be disabled but the service has old data in it, we still don't want to use that  data (but do preserve it)

maplain replied:
added gate check logic before `apiservice.RequestsOnlyLocalTrafficForInternal(service)`

thockin commented on pkg/proxy/topology.go:
```diff
@@ -69,12 +69,38 @@ func FilterTopologyEndpoint(nodeLabels map[string]string, topologyKeys []string,
+// FilterLocalEndpoint should never be called when InternalTrafficPolicy is All.
+// If it is, all endpoints will be directly returned
+func FilterLocalEndpoint(internalTrafficPolicy *v1.ServiceInternalTrafficPolicyType, endpoints []Endpoint) []Endpoint {
+	if internalTrafficPolicy == nil || *internalTrafficPolicy == v1.ServiceInternalTrafficPolicyCluster {
```
need to check gate?

maplain replied:
added logic to return the original endpoints when gate is not on.

thockin commented on pkg/proxy/topology_test.go:
```diff
@@ -476,3 +476,75 @@ func TestFilterTopologyEndpoint(t *testing.T) {
+		name                  string
+		internalTrafficPolicy *v1.ServiceInternalTrafficPolicyType
+		endpoints             []Endpoint
+		expected              []Endpoint
```
YOu'll want to set the gate in this test and carry a bool per test-case

maplain replied:
done

kcmartin commented:
Hello! I am from the Bug Triage team! Friendly reminder that code freeze is starting March 9th, 2021 (less than 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.

As the PR is tagged for 1.21, is it still planned for this release?

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -2933,3 +2936,253 @@ func TestProxierDeleteNodePortStaleUDP(t *testing.T) {
+`,
+		},
+		{
+			name:                  "internalTrafficPolicy is local and there is non-zero local endpoint",
```
this case should ignore the value "local" and act like "cluster"

maplain replied:
make this testcase and `cluster` share the same expectedIPTables string to emphasize this fact

thockin reviewed (commented):
Looking good

Please make sure the gate is being checked in all the right places and that the tests run with and without the gate.  When the gate is disabled, the value should be ignored, even if it is non-nil.  When the gate is enable, there should always be a value (no special handling for nil or "")

maplain commented on pkg/apis/core/types.go:
```diff
@@ -3684,6 +3697,14 @@ type ServiceSpec struct {
 	ExternalTrafficPolicy ServiceExternalTrafficPolicyType
 
+	// InternalTrafficPolicy specifies if the cluster internal traffic
+	// should be routed to all endpoints or node-local endpoints only
```
updated

maplain commented on pkg/apis/core/types.go:
```diff
@@ -3684,6 +3697,14 @@ type ServiceSpec struct {
+	// "Cluster" routes internal traffic to a Service to all endpoints.
+	// "Local" routes traffic to node-local endpoints only, traffic is
+	// dropped if no node-local endpoints are ready.
+	// +optional
```
added `featureGate=<gatename>` to be aligned with `LoadBalancerClass`. Please advise if I should change both to `+featureGate`

maplain replied:
updated to `+featureGate` for both

maplain commented on pkg/apis/core/validation/validation.go:
```diff
@@ -4432,6 +4438,18 @@ func validateServiceExternalTrafficFieldsValue(service *core.Service) field.Erro
+// spec have legal value.
+func validateServiceInternalTrafficFieldsValue(service *core.Service) field.ErrorList {
+	allErrs := field.ErrorList{}
+
```
added the check on feature gate. Also added a testcase for this scenario.

Commit:
Rename fields and methods for BaseServiceInfo

Fields:
1. rename onlyNodeLocalEndpoints to nodeLocalExternal;
2. rename onlyNodeLocalEndpointsForInternal to nodeLocalInternal;
Methods:
1. rename OnlyNodeLocalEndpoints to NodeLocalExternal;
2. rename OnlyNodeLocalEndpointsForInternal to NodeLocalInternal;

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3684,7 +3697,6 @@ type ServiceSpec struct {
 	// +optional
 	ExternalTrafficPolicy ServiceExternalTrafficPolicyType
 
-	// healthCheckNodePort specifies the healthcheck nodePort for the service.
```
why did you remove this line?

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3742,6 +3754,15 @@ type ServiceSpec struct {
+	// "Cluster" routes internal traffic to a Service to all endpoints.
+	// "Local" routes traffic to node-local endpoints only, traffic is
+	// dropped if no node-local endpoints are ready.
+	// featureGate=ServiceInternalTrafficPolicy
```
`+featureGate=`

thockin commented on pkg/apis/core/types.go:
```diff
@@ -3739,9 +3752,18 @@ type ServiceSpec struct {
+	// should be routed to all endpoints or node-local endpoints only.
+	// "Cluster" routes internal traffic to a Service to all endpoints.
+	// "Local" routes traffic to node-local endpoints only, traffic is
+	// dropped if no node-local endpoints are ready.
```
`The default value is "Cluster".`

maplain replied:
updated

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -4272,9 +4285,18 @@ type ServiceSpec struct {
+	// should be routed to all endpoints or node-local endpoints only.
+	// "Cluster" routes internal traffic to a Service to all endpoints.
+	// "Local" routes traffic to node-local endpoints only, traffic is
+	// dropped if no node-local endpoints are ready.
```
As above, spec the default

maplain replied:
updated. also changed `protobuf:"bytes,21` to `protobuf:"bytes,22` because rebase pulls in `LoadBalancerClass` which claims `21`

thockin replied:
thansk!

thockin reviewed (commented):
I have nothing left to pick on.

/approve
/lgtm

maplain commented:
/test pull-kubernetes-unit
/test pull-kubernetes-verify

Commit:
Update generated files

1. `git clean -fxd`
2. `make update`

maplain commented:
Rebase and re-generate static files:
```
    1. `git clean -fxd`
    2. `make update`
```

maplain commented:
weird. it keeps saying `/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh`, but I've tried a few times, and nothing changes..

maplain commented:
/test pull-kubernetes-verify

thockin commented:
Needs rebase

JornShen commented:
for failed check pull-kubernetes-verify, need to run hack/update-openapi-spec.sh

maplain commented:
@JornShen I've tried to run it a few times, but it doesn't update anything locally.. I actually run `FORCE_ALL=true make update` every time before I push to make sure it is in sync.

let me try it one more time

maplain commented:
```
➜  kubernetes git:(internal-traffic-policy) hack/update-openapi-spec.sh
+++ [0307 18:48:44] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/prerelease-lifecycle-gen
Generating prerelease lifecycle code for       26 targets
+++ [0307 18:48:51] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/deepcopy-gen
Generating deepcopy code for      226 targets
+++ [0307 18:49:03] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/defaulter-gen
Generating defaulter code for       88 targets
+++ [0307 18:49:14] Building go targets for darwin/amd64:
    ./vendor/k8s.io/code-generator/cmd/conversion-gen
Generating conversion code for      123 targets
+++ [0307 18:49:37] Building go targets for darwin/amd64:
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
Generating openapi code for KUBE
Generating openapi code for AGGREGATOR
Generating openapi code for APIEXTENSIONS
^A^KGenerating openapi code for CODEGEN
Generating openapi code for SAMPLEAPISERVER
+++ [0307 18:49:52] Building go targets for darwin/amd64:
    ./vendor/github.com/go-bindata/go-bindata/go-bindata
+++ [0307 18:49:54] Building go targets for darwin/amd64:
    cmd/kube-apiserver
etcd --advertise-client-urls http://127.0.0.1:2379 --data-dir /var/folders/l5/66tl8_29065g0jrk7t92qdjm0000gn/T/tmp.vlrfbPAq --listen-client-urls http://127.0.0.1:2379 --log-level=debug > "/dev/null" 2>/dev/null
Waiting for etcd to come up.
+++ [0307 18:51:25] On try 3, etcd: : {"health":"true"}
{"header":{"cluster_id":"14841639068965178418","member_id":"10276657743932975437","revision":"2","raft_term":"2"}}+++ [0307 18:51:25] Starting kube-apiserver
+++ [0307 18:51:34] On try 9, apiserver: : ok
+++ [0307 18:51:34] Updating
    hack/../api/openapi-spec
+++ [0307 18:51:35] SUCCESS
+++ [0307 18:52:35] Clean up complete


➜  kubernetes git:(internal-traffic-policy) gst
On branch internal-traffic-policy
Your branch is up to date with 'my/internal-traffic-policy'.

nothing to commit, working tree clean
```

@JornShen is it possible that this is related to the go version?
```
go version go1.16 darwin/amd64
```

maplain commented:
/test pull-kubernetes-e2e-gce-alpha-features

maplain commented:
/test pull-kubernetes-e2e-gce-alpha-features

maplain commented:
/test pull-kubernetes-verify

JornShen commented:
@maplain  using linux robot will be ok.

https://github.com/JornShen/kubernetes/commit/c1a7b78a38b79cf071df02d858e5e25074f86838

I have upload it to my repo. You can cherry-pick it, maybe.

andrewsykim commented:
Go version seems right, I think we're on 1.16

thockin commented:
/lgtm

maplain commented:
Thanks @thockin for approval, and @andrewsykim and @JornShen for your help!

I manually copied the file attached by Shen and it passed the test. 

Though after copying the file and running `hack/update-openapi-spec.sh` locally again, I do see changes this time, which is clearly wrong:
```
diff --git a/api/openapi-spec/swagger.json b/api/openapi-spec/swagger.json
index 39807afb663..ec45c84199a 100644
--- a/api/openapi-spec/swagger.json
+++ b/api/openapi-spec/swagger.json
@@ -11795,6 +11795,552 @@
       },
       "type": "object"
     },
+    "io.k8s.api.flowcontrol.v1alpha1.FlowDistinguisherMethod": {
+      "description": "FlowDistinguisherMethod specifies the method of a flow distinguisher.",
+      "properties": {
+        "type": {
+          "description": "`type` is the type of flow distinguisher method The supported types are \"ByUser\" and \"ByNamespace\". Required.",
+          "type": "string"
+        }
+      },
+      "required": [
+        "type"
+      ],
+      "type": "object"
+    },
+    "io.k8s.api.flowcontrol.v1alpha1.FlowSchema": {
+      "description": "FlowSchema defines the schema of a group of flows. Note that a flow is made up of a set of inbound API requests with similar attributes and is identified by a pair of strings: the name of the FlowSchema and a \"flow distinguisher\".",
```
